### PR TITLE
[docs] [C++20] [Modules] Ideas for transitioning to modules

### PR DESCRIPTION
**Pull request by ChuanqiXu9**
This patch tries to provide some ideas to transform an existing libraries to modules. I feel this is helpful for users who is interested in modules from my observation.  While the syntax of modules look easy to understand, the practice gets harder if the users want to make compatible work with headers. Especially the `std` module can be installed in clang18, I think such document may be helpful.

I tried to not be too wordy in this document and I don't want the users to have impressions that they have to follow this or this is the best practice. So I tried to use the term `idea` imply the users this is not compulsory.

I add some regular reviewers for modules, but review opinions from users are highly recommended.

I want to land this before the releasing of clang18 (in the early March). We can and should improve this continuously.

### CODE REVIEW

File: clang/docs/StandardCPlusPlusModules.rst [hunk 1]
```diff
@@ -610,6 +610,345 @@ the following style significantly:
...
+-------------------------------
+
+For new libraries, we encourage them to use modules completely from day one if possible.
+This will be pretty helpful to make the whole ecosystems to get ready.
+
+For many existing libraries, it may be a breaking change to refactor themselves
+into modules completely. So that many existing libraries need to provide headers and module
+interfaces for a while to not break existing users.
+Here we provide some ideas to ease the transition process for existing libraries.
+**Note that the this section is only about helping ideas instead of requirement from clang**.
```
Comment (dwblaikie) [inline clang/docs/StandardCPlusPlusModules.rst:623]:
```suggestion
**Note that the this section is only about helpful ideas instead of requirement from clang**.
```

File: clang/docs/StandardCPlusPlusModules.rst [hunk 2]
```diff
@@ -610,6 +610,345 @@ the following style significantly:
 
 The key part of the tip is to reduce the duplications from the text includes.
 
+Ideas for converting to modules
+-------------------------------
+
+For new libraries, we encourage them to use modules completely from day one if possible.
+This will be pretty helpful to make the whole ecosystems to get ready.
+
+For many existing libraries, it may be a breaking change to refactor themselves
```
Comment (dwblaikie) [inline clang/docs/StandardCPlusPlusModules.rst:619]:
```suggestion
For many existing libraries, it may be a breaking change to refactor them
```

File: clang/docs/StandardCPlusPlusModules.rst [hunk 3]
```diff
@@ -610,6 +610,345 @@ the following style significantly:
...
+into modules completely. So that many existing libraries need to provide headers and module
```
Comment (dwblaikie) [inline clang/docs/StandardCPlusPlusModules.rst:620]:
```suggestion
into modules completely. So many existing libraries will need to provide headers and module
```

File: clang/docs/StandardCPlusPlusModules.rst [hunk 4]
```diff
@@ -610,6 +610,345 @@ the following style significantly:
...
+  export module your_library;
+  export namespace your_namespace {
+    using decl_1;
+    using decl_2;
+    ...
+    using decl_n;
+  }
+
+As the example shows, you need to include all the headers containing declarations needs
+to be exported and `using` such declarations in an `export` block. Then, basically,
```
Comment (dwblaikie) [inline clang/docs/StandardCPlusPlusModules.rst:650]:
```suggestion
As the example shows, you need to include all the headers containing declarations that need
to be exported and `using` such declarations in an `export` block. Then, basically,
```

File: clang/docs/StandardCPlusPlusModules.rst [hunk 5]
```diff
@@ -610,6 +610,345 @@ the following style significantly:
...
+  + #ifndef IN_MODULE_INTERFACE
+  #include "third_party/A/headers.h"
+  + #endif
+
+  #include "header_x.h"
+
+  ...
+
+This may be helpful to get better diagnostic messages if you forgot to update your module 
+interface unit file during maintaining.
```
Comment (dwblaikie) [inline clang/docs/StandardCPlusPlusModules.rst:697]:
```suggestion
interface unit file to match changes to the headers in the future.
```

File: clang/docs/StandardCPlusPlusModules.rst [hunk 6]
```diff
@@ -610,6 +610,345 @@ the following style significantly:
...
+to other styles.
+
+ABI breaking style
+~~~~~~~~~~~~~~~~~~
+
+The term ``ABI breaking`` sounds terrifying generally. But you may want it here if you want
+to force your users to introduce your library in a consistent way. E.g., they either include
+your headers all the way or import your modules all the way.
+The style prevents the users to include your headers and import your modules at the same time
+in the same repo.
```
Comment (dwblaikie) [inline clang/docs/StandardCPlusPlusModules.rst:713]:
```suggestion
The style prevents the users from including your headers and importing your modules at the same time
in the same repo.
```

Comment (dwblaikie) [inline clang/docs/StandardCPlusPlusModules.rst:713]: I would imagine this is generally an anti-goal for most people, or would create maintenance challenges - because if your library is used by other libraries - some of those users might be using C++20 modules, and some might not be. And then if you want to depend on all those libraries together in some other library/program, it'd be unfortunate if you were stuck/not able to do that due to the very opinionated original library that said "only use headers or only use modules, don't mix and match".

File: clang/docs/StandardCPlusPlusModules.rst [hunk 7]
```diff
@@ -610,6 +610,345 @@ the following style significantly:
...
+    using third_party_decl_used_in_cpp_2;
+    ...
+    using third_party_decl_used_in_cpp_n;
+  }
+  #endif
+
+(And add `EXPORT` and conditional include to the headers as suggested in the export
+extern-C++ style section)
+
+Remember that the ABI get changed and we need to compile our source files into the
```
Comment (dwblaikie) [inline clang/docs/StandardCPlusPlusModules.rst:751]:
```suggestion
Remember that the ABI is changed and we need to compile our source files into the
```

File: clang/docs/StandardCPlusPlusModules.rst [hunk 8]
```diff
@@ -610,6 +610,345 @@ the following style significantly:
...
+  // used in the .cpp files.
+  namespace third_party_namespace {
+    using third_party_decl_used_in_cpp_1;
+    using third_party_decl_used_in_cpp_2;
+    ...
+    using third_party_decl_used_in_cpp_n;
+  }
+  #endif
+
+In case the number of your source files are small, we may put everything in the private
```
Comment (dwblaikie) [inline clang/docs/StandardCPlusPlusModules.rst:773]:
```suggestion
If you have relatively few source files, you may put everything in the private
```

File: clang/docs/StandardCPlusPlusModules.rst [hunk 9]
```diff
@@ -610,6 +610,345 @@ the following style significantly:
...
+module fragment directly. (it is suggested to add conditional include to the source
+files too). But it will make the compilation of the module interface unit to be slow
+when the number of the source files are not small enough.
```
Comment (dwblaikie) [inline clang/docs/StandardCPlusPlusModules.rst:776]:
```suggestion
files too). But it will make the compilation of the module interface unit slow
if there are many source files.
```

File: clang/docs/StandardCPlusPlusModules.rst [hunk 10]
```diff
@@ -610,6 +610,345 @@ the following style significantly:
...
+
+  + module your_library;
+
+  // Following off should be unchanged.
+  ...
+
+The module implementation unit will import the primary module implicitly.
+We don't include any headers in the module implementation units
+here since we want to avoid duplicated declarations between translation units.
+This is the reason why we add non-exported using declarations from the third
```
Comment (dwblaikie) [inline clang/docs/StandardCPlusPlusModules.rst:801]:
```suggestion
This is the reason why we add non-exported using declarations from third
```

File: clang/docs/StandardCPlusPlusModules.rst [hunk 11]
```diff
@@ -610,6 +610,345 @@ the following style significantly:
...
+provide a modular one ``libyour_library_modules.so`` since you changed the ABI.
+
+What if there are headers only inclued by the source files
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The above practice may be problematic if there are headers only included by the source
+files. If you're using private module fragment, you may solve the issue by including them
+in the private module fragment. While it is OK to solve it by including the implementation
+headers in the module purview if you're using implementation module units, it may be
+suboptimal since the primary module interface units now containing entities not belongs
```
Comment (dwblaikie) [inline clang/docs/StandardCPlusPlusModules.rst:814]:
```suggestion
suboptimal since the primary module interface units now contains entities not belongs
```

File: clang/docs/StandardCPlusPlusModules.rst [hunk 12]
```diff
@@ -610,6 +610,345 @@ the following style significantly:
...
+
+We don't show code snippet since it may be too verbose or not good or not general.
+But it may not be too hard if you can understand the points of the section.
+
+Providing a header to skip parsing redundant headers
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+It is a problem for clang to handle redeclarations between translation units.
+Also there is a long standing issue in clang (`problematic include after import <https://github.com/llvm/llvm-project/issues/61465>`_).
+But even if the issue get fixed in clang someday, the users may still get slower compilation speed
```
Comment (dwblaikie) [inline clang/docs/StandardCPlusPlusModules.rst:830]:
```suggestion
But even if the issue gets fixed in clang someday, the users may still get slower compilation speed
```

**Review by dwblaikie — COMMENTED** [review]
Generally seems like good stuff to write down - I'm not sure about the ABI breaking version (that mostly seems like it would make more problems than solutions).

Commented on some minor grammatical issues.

File: clang/docs/StandardCPlusPlusModules.rst (hunk 6 above)
Comment (ChuanqiXu9) [reply clang/docs/StandardCPlusPlusModules.rst]: Personally I'd like to do this for my library. I feel the idea is somewhat like `private` or `final` in C++. They make the project to be clean by restricting the usages. I know this is not universally true. For example, fundamental libraries like boost or STL may not be in this. 

And I feel this may be fine since it is only an idea instead of requiring  people to follow. I just want to give a hint to people who want to do this but don't know how.

**Comment by ChuanqiXu9** [comment]
Thanks for the reviewing.

File: clang/docs/StandardCPlusPlusModules.rst (hunk 6 above)
Comment (dwblaikie) [reply clang/docs/StandardCPlusPlusModules.rst]: What's the motivation for doing this? Given the challenges it presents to downstream usage of the library?

**Comment by ChuanqiXu9** [comment]
I'd like to land this in 2 weeks if no more comments come in.

**Review by dwblaikie — APPROVED** [review]
Seems ok
